Fixed get ac_id

D-VR opened the pull request:
10.0.0.55 will always return ac_id = 1,
even when in reality it's not the correct one, leading to failure to login in some instances.

10.0.0.53 always returns the correct ac_id

Commit:
Fixed get ac_id

10.0.0.55 will always return ac_id = 1,
even when in reality it's not the correct one, leading to failure to login in some instances.

10.0.0.53 always returns the correct ac_id

phy25 reviewed (changes requested):
`10.0.0.53` only grants access to Intranet because it redirects users to www.bit.edu.cn. Just tested with Ethernet access, and it doesn't work.

D-VR commented:
you are right, i just checked again over Bit Mobile. seems only BIT Web will redirect in that case. I’ll do some more research 

current method is still broken for some BIT Web logins due to wrong ac_id though

D-VR commented:
Bit Web has different behaviour than Ethernet/Bit Mobile. Bit Web will force redirect every url to the correct ac_id, bit mobile and ethernet default to ac_id=1 and do not redirect, this makes it a bit difficult.

maybe check for redirect, and if not default to ac_id=1?

phy25 commented:
It's best to use a 204 response to an Internet URL to check for that: https://www.v2ex.com/t/189687

D-VR commented:
>It's best to use a 204 response to an Internet URL to check for that: https://www.v2ex.com/t/189687

@phy25

i was thinking of using a captive portal detection url like:
captive.apple.com
or
www.msftconnecttest.com/connecttest.txt
or
detectportal.firefox.com/success.txt
~~(this one seems to be blocked for some reason)~~ unreliable

Commit:
Real Fix for ac_id

proposed fix for ac_id problem.

havent touched urllib, because its a mess to implement using exceptions, and i'm not familiar with it.

D-VR commented:
This ac_id method, works on both BIT Web and BIT Mobile, i tried changing py2 using the same method, but urllib wasnt playing nice with me, and i'm not too familiar with it.
Tell me what you think.
Logic:

get http://detectportal.firefox.com/success.txt
if no response in 0.01seconds
set ac_id = 1 
#detection if on BIT mobile or Ethernet

else if response (redirect)
set ac_id = according to redirect
#detection for BIT Web

else if response (success)
ac_id = 1 
#incase if already logged in AND site loads (which it doesnt most of the time) so script doesnt break

Commit:
Accidental Comment

phy25 commented:
In my opinion use an address in China mainland is much better (see the v2ex link attached before). And I don't think the request to detectportal.firefox.com (Akamai CDN Japan) would finish in 10ms, maybe 1000ms is better?

D-VR commented:
>In my opinion use an address in China mainland is much better (see the v2ex link attached before). And I don't think the request to detectportal.firefox.com (Akamai CDN Japan) would finish in 10ms, maybe 1000ms is better?

we don’t want to get an answer to the request though. since if it connects, that means we are already online, and we don’t have to login in the first place. 

(we could maybe use a non existing address?)

what we want is the redirect, which happens on the internal network so response time should be less than 10 ms (might increase it, depending on testing, but was sufficient in my original run)

(note, the timeout is not for a finished response, but for first received packet from server)

phy25 commented:
Makes sense, but I still stand for 0.05s 🙂, because some WiFi may be slower.

D-VR commented:
>Makes sense, but I still stand for 0.05s 🙂, because some WiFi may be slower.

@phy25

sounds good, also any idea how to implement this in urllib?

RogerYong commented:
~~How about [http://10.0.0.55/index_8.html](http://10.0.0.55/index_8.html)?~~

D-VR commented:
does 8 work for BIT mobile and ethernet?

D-VR commented:
one more thing, increasing timeout like @phy25 recommended might be a good idea, not sure how stable wifi authentication is 

(also the scripts using urllib still use old method.. not sure how many routers will be connecting over BIT Web though, so fix might be unnecessary)

D-VR commented:
@RogerYong ~~50~~100~80ms is advised, just had the problem where wifi was too slow resulting in wrong login using ac_id=1

btw, using the wrong ac_id, will result in you logging in, but without being able to access the internet, in addition, if you go to the correct ac_id and login (while still being logged in as the wrong ac_id) you will not be able to login, until you log off the first ac_id...

~~pretty sure it’s possible to log out with an ac_id = null~~ it logs you out. completely. even from 10.0.0.55. not able to access any ip afterwards. possible bug with srun?
